File: app/src/main/java/pl/mareklangiewicz/myhub/MyReposPresenter.java
package pl.mareklangiewicz.myhub;

import android.support.annotation.MainThread;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.noveogroup.android.log.MyLogger;

import java.util.Locale;

import javax.inject.Inject;

import pl.mareklangiewicz.myhub.data.Account;
import pl.mareklangiewicz.myhub.mvp.IMyReposView;
import pl.mareklangiewicz.myhub.mvp.Presenter;
import rx.Observer;
import rx.Subscription;

@MainThread
public class MyReposPresenter extends Presenter<IMyReposView> {

    private MyLogger log = MyLogger.UIL;

    private @NonNull MGModel mModel;
    private @Nullable Subscription subscription;

    @Inject MyReposPresenter(@NonNull MGModel model) {
        mModel = model;
    }

    @Override public void attachIView(@NonNull IMyReposView iview) {
        super.attachIView(iview);
        subscription = mModel.loadLatestAccount()
                .subscribe(new Observer<Account>() {
                    @Override public void onCompleted() {
                        log.v("loading completed.");
                    }

                    @Override public void onError(Throwable e) {
                        log.e(e, "[SNACK]Error %s", e.getLocalizedMessage());
                    }

                    @Override public void onNext(Account account) {
                        showAccount(account);
                    }
                });
    }

    @Override public void detachIView() {
        if(subscription != null && !subscription.isUnsubscribed())
            subscription.unsubscribe();
        subscription = null;
        super.detachIView();
    }

    /**
     * Displays given account on attached IView. Clears IView if account is null.
     */
    private void showAccount(@Nullable Account account) {
        IMyReposView view = getIView();
        if(view != null) {
            view.setStatus(
                    account != null
                            ?
                            String.format(Locale.US, "%s: %d repos (%tF %tT):",
                                    account.getLogin(),
                                    account.getRepos().size(),
                                    account.getTime(),
                                    account.getTime()
                            )
                            :
                            "no repos loaded"
            );
            view.setRepos(account != null ? account.getRepos() : null);
        }

    }
}

File: app/src/main/java/pl/mareklangiewicz/myhub/MyReposPresenter.kt
package pl.mareklangiewicz.myhub

import android.support.annotation.MainThread
import io.reactivex.disposables.Disposable
import io.reactivex.disposables.Disposables
import pl.mareklangiewicz.myhub.data.Account
import pl.mareklangiewicz.myhub.mvp.IMyReposDiew
import pl.mareklangiewicz.myhub.mvp.IPresenter
import pl.mareklangiewicz.myloggers.MyAndroLogger
import pl.mareklangiewicz.myutils.Cancel
import pl.mareklangiewicz.myutils.Lst
import pl.mareklangiewicz.myutils.asLst
import pl.mareklangiewicz.myutils.lsubscribe
import javax.inject.Inject
import javax.inject.Named

@MainThread
class MyReposPresenter @Inject constructor(private val model: MHModel, @Named("UI") private val log: MyAndroLogger)
: IPresenter<IMyReposDiew> {

    private var clicksSub: (Cancel) -> Unit = { }
    private var loadLatestAccountSubscription: Disposable = Disposables.disposed()

    override var xiew: IMyReposDiew? = null

        get() = field

        set(value) {

            clicksSub(Cancel)
            loadLatestAccountSubscription.dispose()

            field = value
            if (value == null) return

            loadLatestAccountSubscription = model.loadLatestAccount().lsubscribe(log, logOnCompleted = "load latest account repos completed") {
                showAccount(it)
            }

            value.progress.visible = false // we don't use it at all.. (currently)
        }

    /** Displays given account on attached IView. Clears IView if account is null. */
    private fun showAccount(account: Account?) {

        val v = xiew ?: return

        v.data = account

        clicksSub(Cancel)

        clicksSub = v.repos.clicksFromRepos {
            v.notes.data = it.data?.notes?.asLst() ?: Lst()
            v.showNotes()
        }
    }
}
